pgml-cms/blog/our-migration-from-aws-to-gcp-with-minimal-downtime.md

```diff
@@ -0,0 +1,134 @@
+
+Once the ZFS snapshot finally made it from the East coast to the Midwest, we configured Postgres streaming replication, which went as you'd expect, and we had a live hot standby in GCP, ready to go. Before cutting the AWS cord, we wanted to double check that everything was okay. We were moving customer data after all, and losing data is bad for business — especially for a database company.
+
+#### The case of the missing bytes
+
+ZFS is a reliable and battle tested filesystem, so we were not worried, but there is nothing wrong with a second opinion. The naive way to check that all your data is still there is to compare the size of the filesystems. Not a terrible place to start, so we ran `df -h` and immediately our jaws dropped: only half the data made it over to GCP.
+
+After days of roadblocks, this was not a good sign, and there was no reasonable explanation for what happened. ZFS checksums every single block, mbuffer is a simple tool, pbzip definitely decompressed the stream and SSH hasn't lost a byte since the 1990s.
+
+In addition, just to make things even weirder, Postgres replication didn't complain and the data was, seemingly, all there. We checked by running your typical `SELECT COUNT(*) FROM a_few_tables` and everything added up: as the data was changing in AWS, it was updating in GCP.
+
+#### (File)systems are virtual
+
+If you ever tried to find out how much free memory your computer has, you know there is no obvious answer. Are you asking for RSS of every single process, virtual memory, and do you have swap enabled, and are you considering the kernel page cache or fragmentation? At the end, you just have to trust that the kernel knows what it is doing.
+
+Filesystems are exactly the same, and to the uninitiated, the difference in file sizes can be scary. After a few Google searches and reading a bunch of panicked system administrator's forum posts from the mid-2000s, it was the manual page for `du(1)` that provided the answer:
+
+```
+--apparent-size
+    print  apparent  sizes, rather than disk usage; although the apparent size is usually smaller, it may be
+    larger due to holes in ('sparse') files, internal fragmentation, indirect blocks, and the like
+```
+
+The database files were the same on GCP and AWS, if one checked them for their "apparent" size: the size of the file as seen by applications, not what they actually used on disk. ZFS is quite clever, and during the transfer with `zfs send`, repacked the filesystem which was somewhat fragmented after years of random writes.
+
+### The cutover
+
+The final step was to move our customers' traffic from AWS to GCP, and do so without losing a byte of data. We picked the lowest traffic period, midnight Pacific time, and shut down our AWS primary.
+
+As soon as the Systemd service stopped, we changed the DNS record to point to our GCP standby and ran `SELECT pg_promote()`. Traffic moved over almost immediately, thanks to our low DNS TTL and we were back in business.
```

Comment:
It might be worth mentioning 0 downtime pgcat cutovers as a possibility for _in-datacenter_ machine transfers, but how that isn't an option when you're moving pgcat poolers themselves to stay close to the database and minimize latency.